pkg/cmd/pulumi/state_edit_frontend.go

```diff
@@ -0,0 +1,480 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"sort"
+	"strconv"
+	"strings"
+	"sync"
+
+	dyff "github.com/dixler/dyff/pkg/pulumi"
+	"github.com/pulumi/pulumi/pkg/v3/backend"
+	"github.com/pulumi/pulumi/pkg/v3/backend/httpstate"
+	"github.com/pulumi/pulumi/pkg/v3/backend/httpstate/client"
+	"github.com/pulumi/pulumi/pkg/v3/resource/deploy"
+	"github.com/pulumi/pulumi/pkg/v3/resource/stack"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/apitype"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/workspace"
+	yaml "gopkg.in/yaml.v3"
+)
+
+type stateFrontend interface {
+	GetBackingFile() string
+	Reset() error
+	SaveToFile(state apitype.DeploymentV3) error
+	ToSnapshot() (deploy.Snapshot, error)
+	Diff() (string, error)
+}
+
+type jsonStateFrontend struct {
+	backingFile string
+	original    *apitype.DeploymentV3
+}
+
+var _ stateFrontend = &jsonStateFrontend{}
+
+func (se *jsonStateFrontend) GetBackingFile() string {
+	return ".pulumi-state-edit.json"
+}
+
+func (se *jsonStateFrontend) Reset() error {
+	contract.Assertf(se.original != nil, "original must not be nil")
+	return se.SaveToFile(*se.original)
+}
+
+func (se *jsonStateFrontend) SaveToFile(state apitype.DeploymentV3) error {
+	if se.original == nil {
+		se.original = &state
+	}
+	if se.backingFile == "" {
+		se.backingFile = ".pulumi-state-edit.json"
+	}
+	writer, err := os.Create(se.backingFile)
+	if err != nil {
+		return fmt.Errorf("could not open file: %w", err)
+	}
+	enc := json.NewEncoder(writer)
+	enc.SetEscapeHTML(false)
+	enc.SetIndent("", "    ")
+
+	if err = enc.Encode(state); err != nil {
+		return fmt.Errorf("could not serialize deployment as JSON : %w", err)
+	}
+	return nil
+}
+
+func (se *jsonStateFrontend) ToSnapshot() (deploy.Snapshot, error) {
+	b, err := os.ReadFile(se.backingFile)
+	if err != nil {
+		return deploy.Snapshot{}, err
+	}
+	ctx := context.Background()
+	dep, err := stack.DeserializeUntypedDeployment(ctx, &apitype.UntypedDeployment{
+		Version:    3,
+		Deployment: b,
+	}, stack.DefaultSecretsProvider)
+	if err != nil {
+		return deploy.Snapshot{}, err
+	}
+
+	if dep == nil {
+		return deploy.Snapshot{}, errors.New("could not deserialize deployment")
+	}
+	return *dep, nil
+}
+
+func (se *jsonStateFrontend) Diff() (string, error) {
+	contract.Assertf(se.original != nil, "original must not be nil")
+	old := jsonStateFrontend{backingFile: ".pulumi-state-edit.original.json"}
+	err := old.SaveToFile(*se.original)
+	if err != nil {
+		return "", err
+	}
+
+	diff, err := dyff.Compare(old.backingFile, se.backingFile)
```

Comment:
We should be diff'ing the snapshot, not the textual data. This should be using the same logic as we use for things like a refresh preview.

Reply:
The `pulumi stack export` and `pulumi stack import` commands operate on the state file. I'm aiming to add a simple wrapper for that workflow with some warnings. We have ideation today and can discuss further.

Reply:
> I'm not aware of a straightforward snapshot diff function and would appreciate a link!

Me neither I'm afraid :( display logic is strangely tangled up with engine events. But it feels like the right way to display these results, and I imagine comes in useful elsewhere as well.

We'll discuss at meeting later.